clippy_lints/src/methods/useless_nonzero_new_unchecked.rs

```diff
@@ -0,0 +1,61 @@
+use clippy_utils::diagnostics::{span_lint_and_sugg, span_lint_and_then};
+use clippy_utils::is_in_const_context;
+use clippy_utils::msrvs::{self, Msrv};
+use clippy_utils::source::snippet_with_applicability;
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_errors::Applicability;
+use rustc_hir::{Block, BlockCheckMode, Expr, ExprKind, Node, QPath, UnsafeSource};
+use rustc_hir_analysis::lower_ty;
+use rustc_lint::LateContext;
+use rustc_span::sym;
+
+use super::USELESS_NONZERO_NEW_UNCHECKED;
+
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'_>, func: &Expr<'tcx>, args: &[Expr<'_>], msrv: &Msrv) {
+    if msrv.meets(msrvs::CONST_UNWRAP)
+        && let ExprKind::Path(QPath::TypeRelative(ty, segment)) = func.kind
+        && segment.ident.name == sym::new_unchecked
+        && let [init_arg] = args
+        && let ExprKind::Lit(_) = init_arg.kind
+        && is_in_const_context(cx)
+        && is_type_diagnostic_item(cx, lower_ty(cx.tcx, ty), sym::NonZero)
```

Comment:
Since this is within a body, `lower_ty` can ICE and shouldn't be used (e.g. `NonZero::<_>::new_unchecked(1)` currently does). `cx.typeck_results().node_type(ty.hir_id)` should be fine